### PR TITLE
Implement package data curation

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Please see the individual commit messages for details.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/211)

### CODE REVIEW

**Comment by mnonnenmacher** [comment]
Review status: 35 of 46 files reviewed at latest revision, 9 unresolved discussions.

---

*[analyzer/src/test/assets/package-curations.yml, line 3 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L33-nzjcmdSbESSsa8r:-L335R3E1IDcJnKJI3Vl:bcjwnid) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/f25760df65fd29a64dee376292ee76fba1aaa908/analyzer/src/test/assets/package-curations.yml#L3)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

While looking at this YAL file it occurred to me that simply from looking at its contents, you have now real idea what this `data` is about. Should we rename `data` to `curation` or `curations` in the YAML output?
</blockquote></details>

Done, I was thinking about also renaming it in code for consistency, but there data sounds good to me.

---

*[analyzer/src/test/kotlin/YamlFilePackageCurationProviderTest.kt, line 31 at r3](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L3309L6RzrqfvQgn68G:-L336iLNbiap0Z7qBl2j:bmfxqef) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/f25760df65fd29a64dee376292ee76fba1aaa908/analyzer/src/test/kotlin/YamlFilePackageCurationProviderTest.kt#L31)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

You know what I'm going to nit about, right? :-D
</blockquote></details>

Done...

---

*[model/src/main/kotlin/Identifier.kt, line 73 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32tlh9m8-p5phDvM4m:-L336u6cYHAXFno9uleL:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/16af9c64906af671c77de4f0c15868c73192d476/model/src/main/kotlin/Identifier.kt#L73)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Nit: Comma after "To match".
</blockquote></details>

Done.

---

*[model/src/main/kotlin/Identifier.kt, line 86 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32twzMRrsiDCr1NcHx:-L3376zOYkSOAO57VZ2s:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/16af9c64906af671c77de4f0c15868c73192d476/model/src/main/kotlin/Identifier.kt#L86)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

How about naming the variable `other`, which I believe is more common for such methods. Also `this` is an `identifier` after all.
</blockquote></details>

Done.

---

*[model/src/main/kotlin/Identifier.kt, line 101 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32uJcjSUtr9KfaQvyR:-L3379SjLoubHloM0PBN:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/16af9c64906af671c77de4f0c15868c73192d476/model/src/main/kotlin/Identifier.kt#L101)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Nit: Return type can be omitted.
</blockquote></details>

Done.

---

*[model/src/main/kotlin/PackageCuration.kt, line 41 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32wKb4fh1dC8mV5zEp:-L33AKGnJz-V9vD3XYLF:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/7a317e1c1585d8ae6557ea8539d58ec7925f6340/model/src/main/kotlin/PackageCuration.kt#L41)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Could you please add docs here, too, so the class is fully documented?
</blockquote></details>

Done.

---

*[model/src/main/kotlin/PackageCurationData.kt, line 75 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32zKeX7Cgi0vNdXWMk:-L337afDRdlU8EP_djCU:bycefrk) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/7a317e1c1585d8ae6557ea8539d58ec7925f6340/model/src/main/kotlin/PackageCurationData.kt#L75)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Doesn't this shadow Kotlin's own `apply()` function? How about `patch()` instead? Or `curate()`?
</blockquote></details>

No, the apply from stdlib takes a `T.() -> Unit` argument. I like to say "apply a curation to a package", but can also change it.

---

*[model/src/main/kotlin/PackageCurationData.kt, line 77 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32xNFMEcfFVy3UdSe9:-L338bw_wUVNV0LLzs5P:bv8rphs) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/7a317e1c1585d8ae6557ea8539d58ec7925f6340/model/src/main/kotlin/PackageCurationData.kt#L77)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

Should we now use `vcs.merge(pkg.vcs)` here instead, relying on the current and future implementation of `merge()` to do what we want also here?
</blockquote></details>

I don't think that the semantics for the automatic merge are the same as for the manual curation. E.g. with using `pkg.vcs.merge(curation.vcs)` we could not fix a broken URL, because the existing URL is not blank. And with `curation.vcs.merge(pkg.vcs)` it could happen that the provider is not changed. I think both behaviours are not expected for the curation process.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211)*

**Comment by mnonnenmacher** [comment]
Review status: 35 of 46 files reviewed at latest revision, 9 unresolved discussions.

---

*[scanner/src/main/kotlin/Main.kt, line 213 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32v8HcSXqZs-fqHNlf:-L33AhSlceOXuCr9hc8T:bow797v) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/16af9c64906af671c77de4f0c15868c73192d476/scanner/src/main/kotlin/Main.kt#L213)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

I was first thinking we should change `scanEntry()` to take an `Identifier` now, but as inside `scanEntry()` the `identifier` is only used in string output that's not useful just yet. So I'm fine with keep it as-is.
</blockquote></details>

I was about to change this, but reverted because scanEntry is also called with `inputPath.absolutePath`in line 223.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211)*

**Comment by sschuberth** [comment]
:lgtm:

---

Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[model/src/main/kotlin/PackageCurationData.kt, line 75 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32zKeX7Cgi0vNdXWMk:-L33Icei_vmCI-mxuVDE:b-9sbepg) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/7a317e1c1585d8ae6557ea8539d58ec7925f6340/model/src/main/kotlin/PackageCurationData.kt#L75)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

No, the apply from stdlib takes a `T.() -> Unit` argument. I like to say "apply a curation to a package", but can also change it.
</blockquote></details>

I wasn't really meaning to shadow in the strict sense, but these are now overloads of totally different functions. But we can change that later if it turns out to be annoying.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-:-L33J3TzlQQYpjQmiilU:bnfp4nl)*

File: analyzer/src/main/kotlin/MavenSupport.kt [hunk 1]
```diff
@@ -20,6 +20,7 @@
 package com.here.ort.analyzer
 
 import ch.frankel.slf4k.*
+import com.here.ort.model.Identifier
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/MavenSupport.kt]: Nit: Should belong to the block below.

File: model/src/main/kotlin/Identifier.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+         */
+        @JvmField
+        val EMPTY = Identifier(
+                packageManager = "",
+                namespace = "",
+                name = "",
+                version = ""
+        )
+
+        fun fromString(identifier: String): Identifier {
```
Comment (sschuberth) [inline model/src/main/kotlin/Identifier.kt]: Please also add docs here to have the class fully documented.

File: analyzer/src/main/kotlin/MavenSupport.kt (hunk 1 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/MavenSupport.kt]: Done.

File: model/src/main/kotlin/Identifier.kt (hunk 2 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/Identifier.kt]: Done. I have also changed `matches()` slightly to ignore case when comparing package managers, as we generate those names. For namespace, name, and version I think it is better to not ignore case.

**Comment by mnonnenmacher** [comment]
Review status: 44 of 46 files reviewed at latest revision, 2 unresolved discussions.

---

*[model/src/main/kotlin/PackageCurationData.kt, line 75 at r2](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211#-L32zKeX7Cgi0vNdXWMk:-L33N-t5DaiT_EJOXJcj:b-376tdt) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/7a317e1c1585d8ae6557ea8539d58ec7925f6340/model/src/main/kotlin/PackageCurationData.kt#L75)):*
<details><summary><i>Previously, sschuberth (Sebastian Schuberth) wrote…</i></summary><blockquote>

I wasn't really meaning to shadow in the strict sense, but these are now overloads of totally different functions. But we can change that later if it turns out to be annoying.
</blockquote></details>

Ok, I'll leave the name for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211)*

**Comment by sschuberth** [comment]
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/211)*

**Pull request merged**